packages/itwin/tree-widget/src/components/trees/category-tree/CategoriesTreeComponent.tsx

```diff
@@ -174,14 +174,40 @@ function InvertButton(props: CategoriesTreeHeaderButtonProps) {
 
     const enabled: string[] = [];
     const disabled: string[] = [];
-    for (const id of ids) {
-      if (props.viewport.view.viewsCategory(id)) {
-        enabled.push(id);
+    const enabledSubCategories: string[] = [];
+    const disabledSubCategories: string[] = [];
+
+    (await props.viewport.iModel.categories.getCategoryInfo(ids)).forEach((category) => {
+      if (!props.viewport.view.viewsCategory(category.id)) {
+        disabled.push(category.id);
       } else {
-        disabled.push(id);
+        // First, we need to check if at least one subcategory is disabled. If it is true, then only subcategories should change display, not categories.
+        let isAtLeastOneSubCategoryDisabled: boolean = false;
+        const subCategories = [...category.subCategories.values()];
+        for (const subCategory of subCategories) {
+          if (!props.viewport.isSubCategoryVisible(subCategory.id)) {
+            isAtLeastOneSubCategoryDisabled = true;
+            break;
+          }
+        }
+        if (isAtLeastOneSubCategoryDisabled) {
+          subCategories.forEach((subCategory) => {
+            props.viewport.isSubCategoryVisible(subCategory.id) ? enabledSubCategories.push(subCategory.id) : disabledSubCategories.push(subCategory.id);
+          });
+        } else{
+          enabled.push(category.id);
+        }
       }
```

Comment:
I think this code can be refactored a bit:
1. You can reduce nesting with early returns.
2. use `Array.some` function to check if there are any disabled sub categories.

Reply:
Fixed